File: frontend-old-egui/src/columns.rs
//! Per-column display metadata.
//!
//! Querydown lets the user attach arbitrary metadata to each result column in their
//! query source. The compiler surfaces this as one [`AnnotationValue`] (or `None`) per output
//! column. Here we normalize each blob into a [`ColumnMetadata`] with every field
//! resolved to a concrete value, so rendering never has to think about defaults.

use querydown::AnnotationValue;
use serde::Deserialize;

use crate::format::Formatter;

/// Default column min width, in pixels, when none is specified.
const DEFAULT_MIN_WIDTH: f32 = 0.0;
/// Default column max width, in pixels, when none is specified.
const DEFAULT_MAX_WIDTH: f32 = 500.0;

/// Whether a column's text uses the normal or a de-emphasized ("light") color.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub(crate) enum FontColor {
    #[default]
    Default,
    Light,
}

/// Column text size.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub(crate) enum FontSize {
    #[default]
    Normal,
    Small,
}

/// Horizontal alignment of a column's text within its cell.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub(crate) enum TextAlign {
    #[default]
    Left,
    Right,
    Center,
}

/// Fully-resolved display metadata for a single result column. Every field has a
/// concrete value (defaults already applied), so rendering can use it directly.
#[derive(Debug, Clone, PartialEq)]
pub(crate) struct ColumnMetadata {
    pub(crate) hide: bool,
    pub(crate) min_width: f32,
    pub(crate) max_width: f32,
    pub(crate) font_color: FontColor,
    pub(crate) font_size: FontSize,
    pub(crate) text_align: TextAlign,
    pub(crate) prefix: String,
    pub(crate) suffix: String,
    pub(crate) formatter: Option<Formatter>,
}

impl Default for ColumnMetadata {
    fn default() -> Self {
        Self {
            hide: false,
            min_width: DEFAULT_MIN_WIDTH,
            max_width: DEFAULT_MAX_WIDTH,
            font_color: FontColor::default(),
            font_size: FontSize::default(),
            text_align: TextAlign::default(),
            prefix: String::new(),
            suffix: String::new(),
            formatter: None,
        }
    }
}

/// The user-facing width spec: a plain number, or a one/two-element list.
#[derive(Deserialize)]
#[serde(untagged)]
enum WidthSpec {
    Single(f32),
    List(Vec<f32>),
}

/// The raw metadata blob as written by the user, before defaults are applied. Every
/// field is optional; unknown keys are ignored.
#[derive(Default, Deserialize)]
struct RawColumnMetadata {
    hide: Option<String>,
    width: Option<WidthSpec>,
    color: Option<String>,
    size: Option<String>,
    align: Option<String>,
    prefix: Option<String>,
    suffix: Option<String>,
    /// Kept as a raw JSON value and parsed separately so a malformed formatter blob
    /// yields `None` instead of discarding the whole column's metadata.
    formatter: Option<serde_json::Value>,
}

/// Resolves a [`WidthSpec`] into a `(min_width, max_width)` pair, applying the rules:
/// a plain number sets both bounds equal; `[min]` keeps the default max; `[min, max]`
/// sets both (extra elements ignored); an empty list falls back to the defaults. If the
/// resulting min exceeds the max, the two are swapped.
fn resolve_width(spec: Option<WidthSpec>) -> (f32, f32) {
    let (mut min, mut max) = match spec {
        Some(WidthSpec::Single(n)) => (n, n),
        Some(WidthSpec::List(list)) => match list.as_slice() {
            [] => (DEFAULT_MIN_WIDTH, DEFAULT_MAX_WIDTH),
            [min] => (*min, DEFAULT_MAX_WIDTH),
            [min, max, ..] => (*min, *max),
        },
        None => (DEFAULT_MIN_WIDTH, DEFAULT_MAX_WIDTH),
    };
    if min > max {
        std::mem::swap(&mut min, &mut max);
    }
    (min, max)
}

fn resolve_font_color(value: Option<&str>) -> FontColor {
    match value {
        Some("light") => FontColor::Light,
        _ => FontColor::Default,
    }
}

fn resolve_font_size(value: Option<&str>) -> FontSize {
    match value {
        Some("small") => FontSize::Small,
        _ => FontSize::Normal,
    }
}

fn resolve_text_align(value: Option<&str>) -> TextAlign {
    match value {
        Some("right") => TextAlign::Right,
        Some("center") => TextAlign::Center,
        _ => TextAlign::Left,
    }
}

impl ColumnMetadata {
    /// Normalizes one column's metadata blob (or its absence) into resolved settings. A
    /// `None` blob, or any blob that isn't a JSON object, yields all defaults.
    pub(crate) fn from_meta(meta: Option<&AnnotationValue>) -> Self {
        let raw = meta
            .and_then(|m| serde_json::to_value(m).ok())
            .and_then(|v| serde_json::from_value::<RawColumnMetadata>(v).ok())
            .unwrap_or_default();
        let (min_width, max_width) = resolve_width(raw.width);
        Self {
            hide: raw.hide.as_deref() == Some("yes"),
            min_width,
            max_width,
            font_color: resolve_font_color(raw.color.as_deref()),
            font_size: resolve_font_size(raw.size.as_deref()),
            text_align: resolve_text_align(raw.align.as_deref()),
            prefix: raw.prefix.unwrap_or_default(),
            suffix: raw.suffix.unwrap_or_default(),
            formatter: raw
                .formatter
                .and_then(|v| serde_json::from_value::<Formatter>(v).ok()),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn none_yields_defaults() {
        assert_eq!(ColumnMetadata::from_meta(None), ColumnMetadata::default());
    }

    #[test]
    fn non_object_yields_defaults() {
        let meta = AnnotationValue::String("nonsense".to_string());
        assert_eq!(
            ColumnMetadata::from_meta(Some(&meta)),
            ColumnMetadata::default()
        );
    }

    fn object(entries: Vec<(&str, AnnotationValue)>) -> AnnotationValue {
        AnnotationValue::Object(
            entries
                .into_iter()
                .map(|(k, v)| (k.to_string(), v))
                .collect(),
        )
    }

    fn num(n: &str) -> AnnotationValue {
        AnnotationValue::Number(n.to_string())
    }

    #[test]
    fn parses_all_fields() {
        let meta = object(vec![
            ("hide", AnnotationValue::String("yes".to_string())),
            ("color", AnnotationValue::String("light".to_string())),
            ("size", AnnotationValue::String("small".to_string())),
            ("align", AnnotationValue::String("center".to_string())),
            ("prefix", AnnotationValue::String("$".to_string())),
            ("suffix", AnnotationValue::String(" kg".to_string())),
        ]);
        let m = ColumnMetadata::from_meta(Some(&meta));
        assert!(m.hide);
        assert_eq!(m.font_color, FontColor::Light);
        assert_eq!(m.font_size, FontSize::Small);
        assert_eq!(m.text_align, TextAlign::Center);
        assert_eq!(m.prefix, "$");
        assert_eq!(m.suffix, " kg");
    }

    #[test]
    fn hide_only_yes_hides() {
        let meta = object(vec![("hide", AnnotationValue::String("no".to_string()))]);
        assert!(!ColumnMetadata::from_meta(Some(&meta)).hide);
    }

    #[test]
    fn unknown_enum_values_fall_back_to_defaults() {
        let meta = object(vec![
            ("color", AnnotationValue::String("rainbow".to_string())),
            ("align", AnnotationValue::String("justify".to_string())),
        ]);
        let m = ColumnMetadata::from_meta(Some(&meta));
        assert_eq!(m.font_color, FontColor::Default);
        assert_eq!(m.text_align, TextAlign::Left);
    }

    #[test]
    fn width_plain_number_sets_both_bounds() {
        let meta = object(vec![("width", num("120"))]);
        let m = ColumnMetadata::from_meta(Some(&meta));
        assert_eq!((m.min_width, m.max_width), (120.0, 120.0));
    }

    #[test]
    fn width_single_element_list_keeps_default_max() {
        let meta = object(vec![("width", AnnotationValue::Array(vec![num("80")]))]);
        let m = ColumnMetadata::from_meta(Some(&meta));
        assert_eq!((m.min_width, m.max_width), (80.0, DEFAULT_MAX_WIDTH));
    }

    #[test]
    fn width_two_element_list_sets_both() {
        let meta = object(vec![(
            "width",
            AnnotationValue::Array(vec![num("50"), num("300")]),
        )]);
        let m = ColumnMetadata::from_meta(Some(&meta));
        assert_eq!((m.min_width, m.max_width), (50.0, 300.0));
    }

    #[test]
    fn width_swaps_when_min_exceeds_max() {
        let meta = object(vec![(
            "width",
            AnnotationValue::Array(vec![num("300"), num("50")]),
        )]);
        let m = ColumnMetadata::from_meta(Some(&meta));
        assert_eq!((m.min_width, m.max_width), (50.0, 300.0));
    }

    #[test]
    fn parses_formatter_blob() {
        let meta = object(vec![(
            "formatter",
            object(vec![(
                "type",
                AnnotationValue::String("duration".to_string()),
            )]),
        )]);
        let m = ColumnMetadata::from_meta(Some(&meta));
        assert_eq!(m.formatter, Some(Formatter::Duration {}));
    }

    #[test]
    fn unknown_formatter_type_is_ignored_without_disturbing_other_fields() {
        let meta = object(vec![
            ("align", AnnotationValue::String("right".to_string())),
            (
                "formatter",
                object(vec![("type", AnnotationValue::String("bogus".to_string()))]),
            ),
        ]);
        let m = ColumnMetadata::from_meta(Some(&meta));
        assert_eq!(m.formatter, None);
        assert_eq!(m.text_align, TextAlign::Right);
    }

    #[test]
    fn width_empty_list_uses_defaults() {
        let meta = object(vec![("width", AnnotationValue::Array(vec![]))]);
        let m = ColumnMetadata::from_meta(Some(&meta));
        assert_eq!(
            (m.min_width, m.max_width),
            (DEFAULT_MIN_WIDTH, DEFAULT_MAX_WIDTH)
        );
    }
}

File: frontend/src/query/columns.ts
// Per-column display metadata.
//
// Querydown lets the user attach arbitrary metadata to each result column. The
// compiler surfaces this as one `AnnotationValue` (or `null`) per output column
// (already a JS object — no JSON.parse). Here we normalize each blob into a
// resolved {@link ColumnMetadata} with every field concrete, so rendering never
// has to think about defaults.

import type { AnnotationValue } from "querydown-js";
import { parseFormatter, type Formatter } from "./format";

/** Default column min width, in pixels, when none is specified. */
const DEFAULT_MIN_WIDTH = 0;
/** Default column max width, in pixels, when none is specified. */
const DEFAULT_MAX_WIDTH = 500;

/** Whether a column's text uses the normal or a de-emphasized ("light") color. */
export type FontColor = "default" | "light";
/** Column text size. */
export type FontSize = "normal" | "small";
/** Horizontal alignment of a column's text within its cell. */
export type TextAlign = "left" | "right" | "center";

/** Fully-resolved display metadata for a single result column. Every field has a
 * concrete value (defaults already applied), so rendering can use it directly. */
export interface ColumnMetadata {
  hide: boolean;
  min_width: number;
  max_width: number;
  font_color: FontColor;
  font_size: FontSize;
  text_align: TextAlign;
  prefix: string;
  suffix: string;
  formatter: Formatter | undefined;
}

export function defaultColumnMetadata(): ColumnMetadata {
  return {
    hide: false,
    min_width: DEFAULT_MIN_WIDTH,
    max_width: DEFAULT_MAX_WIDTH,
    font_color: "default",
    font_size: "normal",
    text_align: "left",
    prefix: "",
    suffix: "",
    formatter: undefined,
  };
}

/** Recursively converts a raw annotation value into plain JS. `querydown-js`
 * (serde-wasm-bindgen) serializes Rust map-like `AnnotationValue::Object`s as JS
 * **`Map`s**, not plain objects — so reading `raw.width` by property access would
 * always be `undefined` and every column would fall back to defaults. Normalize
 * `Map`s (and nested ones, e.g. the `formatter` blob) to objects here so the rest
 * of the resolution reads them uniformly; plain objects (the fixture/test path)
 * pass through unchanged. */
function normalizeAnnotation(v: unknown): AnnotationValue | null {
  if (v instanceof Map) {
    const obj: Record<string, AnnotationValue> = {};
    for (const [k, val] of v as Map<unknown, unknown>) {
      obj[String(k)] = normalizeAnnotation(val) as AnnotationValue;
    }
    return obj;
  }
  if (Array.isArray(v)) {
    return v.map((x) => normalizeAnnotation(x) as AnnotationValue);
  }
  if (v !== null && typeof v === "object") {
    const obj: Record<string, AnnotationValue> = {};
    for (const [k, val] of Object.entries(v)) {
      obj[k] = normalizeAnnotation(val) as AnnotationValue;
    }
    return obj;
  }
  return (v ?? null) as AnnotationValue | null;
}

/** Reads a field as a string, or `undefined` when it is absent or not a string. */
function asString(v: AnnotationValue | undefined): string | undefined {
  return typeof v === "string" ? v : undefined;
}

/** Resolves the `width` spec into a `[min, max]` pair: a plain number sets both
 * bounds equal; `[min]` keeps the default max; `[min, max]` sets both (extra
 * elements ignored); an empty list or absent falls back to the defaults. If the
 * resulting min exceeds the max, the two are swapped. */
function resolveWidth(spec: AnnotationValue | undefined): [number, number] {
  let min: number;
  let max: number;
  if (typeof spec === "number") {
    min = spec;
    max = spec;
  } else if (Array.isArray(spec)) {
    const nums = spec.filter((x): x is number => typeof x === "number");
    if (nums.length === 0) {
      min = DEFAULT_MIN_WIDTH;
      max = DEFAULT_MAX_WIDTH;
    } else if (nums.length === 1) {
      min = nums[0];
      max = DEFAULT_MAX_WIDTH;
    } else {
      min = nums[0];
      max = nums[1];
    }
  } else {
    min = DEFAULT_MIN_WIDTH;
    max = DEFAULT_MAX_WIDTH;
  }
  if (min > max) [min, max] = [max, min];
  return [min, max];
}

function resolveFontColor(v: string | undefined): FontColor {
  return v === "light" ? "light" : "default";
}

function resolveFontSize(v: string | undefined): FontSize {
  return v === "small" ? "small" : "normal";
}

function resolveTextAlign(v: string | undefined): TextAlign {
  if (v === "right") return "right";
  if (v === "center") return "center";
  return "left";
}

/** Normalizes one column's annotation blob (or its absence) into resolved
 * settings. A `null` blob, or any blob that isn't a JSON object, yields all
 * defaults. A malformed `formatter` sub-blob yields `undefined` without
 * discarding the rest of the metadata. Ports `ColumnMetadata::from_meta`. */
export function columnMetadataFromAnnotation(
  meta: AnnotationValue | null,
): ColumnMetadata {
  const normalized = normalizeAnnotation(meta);
  if (
    normalized === null ||
    typeof normalized !== "object" ||
    Array.isArray(normalized)
  ) {
    return defaultColumnMetadata();
  }
  const raw = normalized;
  const [minWidth, maxWidth] = resolveWidth(raw.width);
  return {
    hide: asString(raw.hide) === "yes",
    min_width: minWidth,
    max_width: maxWidth,
    font_color: resolveFontColor(asString(raw.color)),
    font_size: resolveFontSize(asString(raw.size)),
    text_align: resolveTextAlign(asString(raw.align)),
    prefix: asString(raw.prefix) ?? "",
    suffix: asString(raw.suffix) ?? "",
    formatter: parseFormatter(raw.formatter),
  };
}
